### PR TITLE
make undefined strict when not capturing macros (#1389)

### LINKED ISSUE #1389: dbt should use jinja's StrictUndefined setting
### Feature description
See the Jinja docs here: http://jinja.pocoo.org/docs/2.10/api/#undefined-types

In dbt versions <= 0.13, a user can invoke an object that hasn't been defined, and it will compile to a falsy empty string. Instead, dbt should raise an exception during compilation, since invoking an undefined object is often the result of a typo or a namespace issue.

For example, the following model file (with a typo `partiion`) compiles and executes:

```
{% set partiion = 'partition by account_id order by deleted_on asc nulls last rows between unbounded preceding and unbounded following' %}
last_value(id) over ({{partition}}) as id,
last_value(address_id) over ({{partition}}) as address_id,
```
Compiles to:
```
last_value(id) over () as id,
last_value(address_id) over () as address_id,
```

### Who will this benefit?
This will benefit any user that sets Jinja objects in their models.

This would be a breaking change for any users intentionally using undefined objects, but I can't think of why that would be desired. Objects that may not be defined could always be checked with `{% if object_name %}`

### PR DESCRIPTION
**Pull request by beckjake**
Fixes #1389 

During the non-macro-capture part of dbt parsing, set the undefined handling to 'strict' - currently we let jinja use its default, which is permissive - many times, overly so.

I expect this will break a good number of macros/models/materializations in end user code.

A fix for a lot of issues is, if `x.y` now raises an error, try `x.get('y', '')`.

### CODE REVIEW

**Comment by drewbanin** [comment]
@beckjake this does look great, but I am just realizing exactly how many macros out in the wild are going to be totally broken once this is merged.

I think this is in the category of "bandaids we should rip off sooner rather than later". Let me play around with this on some example projects to determine the extent of the damage.

**Comment by beckjake** [comment]
We could subclass that class to have it go through a `WARN_ERROR` check, and then users could `--strict` to turn warnings into errors.

**Comment by drewbanin** [comment]
Will that approach arrest parsing? I imagine that once the exception is thrown, dbt isn't going to actually parse the rest of the file, right?

This is broadly what I had in mind, but I was picturing either 1) tucking this thing behind `--strict` mode, or (maybe less desirably) parsing twice, once with `StrictUndefined` to catch any errors (and throwing a warning) and then with the non-strict `Undefined` to actually parse the files. I don't think this is necessarily a good idea, but wanted to throw it out there

**Comment by beckjake** [comment]
I think we're saying the same thing - I'm just saying you'd do something along these lines (untested, but seems reasonable enough):

```
class WarnUndefined(jinja2.StrictUndefined):
    def _fail_with_undefined_warning(self, *args, **kwargs):
        try:
            self._fail_with_undefined_error(*args, **kwargs)
        except Exception as exc:
            if dbt.flags.WARN_ERROR:
                raise
            else:
                logger.warning('WARNING - Undefined jinja: {}'.format(str(exc)))

    __slots__ = ()
    __iter__ = __str__ = __len__ = __nonzero__ = __eq__ = \
    __ne__ = __bool__ = __hash__ = \
        _fail_with_undefined_warning

args['undefined'] = WarnUndefined
```

**Comment by drewbanin** [comment]
Oh, ok, I think i might misunderstand how this jinja handles undefined values. Sure, let's try something like this! I will ping you to make sure I have the right idea :)

**Comment by clrcrl** [comment]
Does this close #1389?

**Comment by drewbanin** [comment]
@clrcrl it does! I updated the description accordingly - thanks! :)

**Comment by drewbanin** [comment]
@beckjake this is super close! Are we able to add a `WARNING` in the event that an undefined value is encountered, but `--warn-error` is not provided? At is stands, there's no indication that dbt encountered an undefined variable unless `--warn-error` is supplied, which doesn't feel right to me.

I think we can use `self._undefined_name` to help print the warning message

**Comment by beckjake** [comment]
>I think we can use self._undefined_name to help print the warning message

So, I definitely could rewrite the function that generates the message and hope it doesn't change in future jinja versions. That said, how much would you object to just doing:

```
try:
    self._fail_with_undefined_error(*args, **kwargs)
except Exception as exc:
    if dbt.flags.WARN_ERROR:
        raise
    else:
        logger.warning('WARNING: {}'.format(str(exc)))
```

**Comment by drewbanin** [comment]
I'm ok with that! Is there a reason we wouldn't go through the existing [warn_or_error](https://github.com/fishtown-analytics/dbt/blob/a4e5a5ac787f16d293aa843872091ebe05eb11c0/core/dbt/exceptions.py#L655) codepath? 

The other thing I would love would be to add info about the resource where the undefined variable was found. Is that something that Jinja would make easy for us? I'm picturing:

```
WARNING: <exception text> in model models/abc.sql
```

Just looking at the code though, I'm not sure how we'd be able to snag that from the `WarnUndefined` scope

**Comment by beckjake** [comment]
>I'm ok with that! Is there a reason we wouldn't go through the existing warn_or_error codepath?

Just that I feel super comfortable letting jinja raise the error and generate a great error message in all sorts of weird edge cases (where does `hint` come from and what kind of stuff can it get up to? I sure don't know!).

**Comment by beckjake** [comment]
If we let jinja do the reporting, I believe we'll catch it [here](https://github.com/fishtown-analytics/dbt/blob/aa4f771df28b307af0cf9fe2fc24432f10a8236b/core/dbt/clients/jinja.py#L264-L271) and display something pleasant that includes the file path, right?

**Comment by drewbanin** [comment]
Ah, I meant specifically for the WARNING case, I assumed we'd decorate the error message for an exception... maybe that's not the case....

Let's get this bad boy merged with just a warning log line. Don't worry about the model path -- we're not doing a good job of handling that anywhere at the moment. Presumably if some var is undefined, it will be pretty easy to cmd+f for :)

**Comment by beckjake** [comment]
>Let's get this bad boy merged with just a warning log line. Don't worry about the model path -- we're not doing a good job of handling that anywhere at the moment. Presumably if some var is undefined, it will be pretty easy to cmd+f for :)

Are you sure? I've come up with a reasonably nice solution for it and it's actually not too hard.

**Comment by drewbanin** [comment]
I take it all back! Yeah, give it a shot then!

**Comment by beckjake** [comment]
On warnings it now looks like this:

```
$ dbt compile
Running with dbt=0.13.0
Found 1 models, 0 tests, 0 archives, 0 analyses, 101 macros, 0 operations, 0 seed files, 0 sources

19:42:39 | Concurrency: 2 threads (target='default')
19:42:39 |
COMPILATION WARNING in model x (models/x.sql)
	'value' is undefined
19:42:40 | Done.
```

File: core/dbt/clients/jinja.py [hunk 1]
```diff
@@ -224,13 +224,81 @@ def __call__(self, *args, **kwargs):
...
+            self.node = node
+
+        def _fail_or_warn_with_undefined_error(self, *args, **kwargs):
+            try:
+                self._fail_with_undefined_error(*args, **kwargs)
+            except Exception as exc:
+                if dbt.flags.WARN_ERROR:
+                    raise
+                else:
+                    msg = 'COMPILATION WARNING in {}:\n\t{}'
```
Comment (drewbanin) [inline core/dbt/clients/jinja.py]:
```suggestion
                    msg = 'Compilation warning in {}:\n\t{}'
```

**Pull request merged**